doc/gearbox/gearbox_mdio-HLD.md

```diff
@@ -0,0 +1,337 @@
+
+### High-Level Design 
+
+The high level design of the gbsyncd mdio access function using the mdio bus from the switch NPU and using single gbsyncd docker in SONiC image can be covered in the following points:
+		
+	- It is an enhancement to the sonic gearbox framework.
+	- Syncd, SAI and some sonic device files are modified.
+	- There are 2 repositories that would be changed, sonic-sairedis and SAI.
+	- gbsyncd will have dependencies on syncd to create the Unix IPC socket. 
+	- Syncd will have new classes for the MDIO IPC server and for the parser of gearbox configuration.
+	- There is a new MDIO IPC client library is added.
+	- The new MDIO IPC client library code can be under syncd, but is not linked to syncd at build time.
+	- External PHY firmware download time always has performance requirements/impact.
+	- For debugging, beside the syslog logging, the Unix socket IPC mechanism can be simulated by socat.
+	- The current change is only tested on the Broadcom switch, but should work for other switch.
+	- SAI switch api has added the mido clause 22 access functions.
+	- The SaiInterface class has added the mdio clause 45/22 read/write virtual functions.
+	- The VendoSai class overrides the mdio clause 45/22 read/write virtual functions.
+	- New keys are added to the gearbox configuration file "gearbox\_config.json".
+	- During the warmboot, the creation of the Unix IPC socket and connection is the same as of coldboot.
+	- The platform module software should not reset the external PHY during warmboot.
+
+The VendorPaiDll class will inherit most member functions from the VendorSai class. The VendorSai class needs some changes to accomdate the inheritance.
+
+When a syncd instance runs inside the syncd docker or inside the gbsyncd docker, the syncd global context number is unique for each instance. We can assume the syncd global context number is 0 for the syncd docker instance and the syncd global context number is larger than 0 for any gbsyncd docker instance.
+
+The syncd global context number can used as an index into the array of the "phys" section in the configuration file "gearbox\_config.json". There is information for PHY id, PAI library name, MDIO library name and MDIO mode stored as values for the JSON key of "phy\_id", "lib\_name", "phy\_access\_lib\_name" and "mdio\_cl22\_only" from the configuration file "gearbox\_config.json".
+
+	int syncd_main(int argc, char **argv)
+	{
+		...
+		if (commandLineOptions->m_globalContext != 0) {
```

Comment:
For the condition `commandLineOptions->m_globalContext != 0`, suggest explicitly define a new command line option such as `m_useVendorPaiDll`. If it is **true**, it will create VendorPaiDll object.

This would be a backward compatible behavior. It does not impact the current case of one gbsyncd process managing multiple same-type PHYs, where it is also a truth of  `m_globalContext != 0`.

Reply:
good point, this was one of the proposal on the meeting, and i think also it is good approach

Reply:
also that json configuration should be passed to that object in the constructor

Reply:
Will add commandLineOptions->m_paiInstance. It severs 2 purposes,  to differentiate VendorPai from VendorSai and to index into the "phys" section array of "gearbox_config.json". The json configuration name is already in commandLineOptions->m_contextConfig which will be passed the VendorPai constructor. The constructor can get all information by parsing the json configuration.